In [4]:
import pandas as pd
import os
import shutil

#IN THIS PROJECT , I HAVE TAKEN A DATASET WHICH COMPRISES OF COVID-19 PATIENTS CHEST XRAYS COLLECTED BY SOME DOCTORS , THE GITHUB LINK IS -'https://github.com/ieee8023/covid-chestxray-dataset'

USING CNN , I WILL CLASSIFY THE XRAY IMAGES INTO COVID +VE AND COVID -VE  

In [2]:
df=pd.read_csv('metadata.csv')

In [3]:
df.head()

patientid  offset sex   age                   finding RT_PCR_positive  \
0         2     0.0   M  65.0  Pneumonia/Viral/COVID-19               Y   
1         2     3.0   M  65.0  Pneumonia/Viral/COVID-19               Y   
2         2     5.0   M  65.0  Pneumonia/Viral/COVID-19               Y   
3         2     6.0   M  65.0  Pneumonia/Viral/COVID-19               Y   
4         4     0.0   F  52.0  Pneumonia/Viral/COVID-19               Y   

  survival intubated intubation_present went_icu  ...              date  \
0        Y         N                  N        N  ...  January 22, 2020   
1        Y         N                  N        N  ...  January 25, 2020   
2        Y         N                  N        N  ...  January 27, 2020   
3        Y         N                  N        N  ...  January 28, 2020   
4      NaN         N                  N        N  ...  January 25, 2020   

                                            location  folder  \
0        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
1        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
2        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
3        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
4  Changhua Christian Hospital, Changhua City, Ta...  images   

                                            filename                   doi  \
0  auntminnie-a-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
1  auntminnie-b-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
2  auntminnie-c-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
3  auntminnie-d-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
4                              nejmc2001573_f1a.jpeg  10.1056/NEJMc2001573   

                                                 url  license  \
0  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
1  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
2  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
3  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
4  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   

                                      clinical_notes other_notes Unnamed: 29  
0  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
1  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
2  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
3  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
4   diffuse infiltrates in the bilateral lower lungs         NaN         NaN  

[5 rows x 30 columns]

In [7]:
img_path='C:/Users/91763/Desktop/COVID-PROJECT/images' 

In [6]:
covid_folder='Dataset/Covid'
if not os.path.exists(covid_folder):
    os.mkdir(covid_folder)
    print('Covid folder created')
else:
    print('There is already a Covid folder')

Covid folder created


In [21]:
for i in range(df.shape[0]):
    df['finding'].values[i]=str( df['finding'].values[i]).split('/')[-1]

In [22]:
df.head()

patientid  offset sex   age   finding RT_PCR_positive survival intubated  \
0         2     0.0   M  65.0  COVID-19               Y        Y         N   
1         2     3.0   M  65.0  COVID-19               Y        Y         N   
2         2     5.0   M  65.0  COVID-19               Y        Y         N   
3         2     6.0   M  65.0  COVID-19               Y        Y         N   
4         4     0.0   F  52.0  COVID-19               Y      NaN         N   

  intubation_present went_icu  ...              date  \
0                  N        N  ...  January 22, 2020   
1                  N        N  ...  January 25, 2020   
2                  N        N  ...  January 27, 2020   
3                  N        N  ...  January 28, 2020   
4                  N        N  ...  January 25, 2020   

                                            location  folder  \
0        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
1        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
2        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
3        Cho Ray Hospital, Ho Chi Minh City, Vietnam  images   
4  Changhua Christian Hospital, Changhua City, Ta...  images   

                                            filename                   doi  \
0  auntminnie-a-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
1  auntminnie-b-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
2  auntminnie-c-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
3  auntminnie-d-2020_01_28_23_51_6665_2020_01_28_...  10.1056/nejmc2001272   
4                              nejmc2001573_f1a.jpeg  10.1056/NEJMc2001573   

                                                 url  license  \
0  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
1  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
2  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
3  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   
4  https://www.nejm.org/doi/full/10.1056/NEJMc200...      NaN   

                                      clinical_notes other_notes Unnamed: 29  
0  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
1  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
2  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
3  On January 22, 2020, a 65-year-old man with a ...         NaN         NaN  
4   diffuse infiltrates in the bilateral lower lungs         NaN         NaN  

[5 rows x 30 columns]

#THERE ARE XRAYS WHICH ARE PRESENT IN A TOP VIEW MANNER BUT WE WANT ONLY THE FRONT VIEW AND ONLY THE COVID-19 PATIENTS

In [31]:
df=df[(df['finding']=='COVID-19')&(df['view']=='PA')]

In [38]:
df.shape

(196, 30)

#MOVING THE COVID-19 XRAY IMAGES TO THE COVID FOLDER WHICH WE HAVE MADE EARLIER

In [41]:
for i in range(df.shape[0]):
    file_name=df['filename'].values[i]
    image_path=os.path.join(img_path,file_name)
    covid_image_path=os.path.join(covid_folder,file_name)
    shutil.copy2(image_path,covid_image_path)

#FETCHING THE NON-COVID XRAYS WHICH WE WILL TAKE IT FROM A KAGGLE DATASET-'https://www.kaggle.com/paultimothymooney/chest-xray-pneumonia'

In [58]:
non_covid_path=r'C:\Users\91763\Downloads\archive (5)\chest_xray\train\NORMAL'

In [59]:
non_covid_folder='Dataset/non-covid'

In [60]:
non_covid=os.listdir(non_covid_path)

In [61]:
import random as rnd

In [62]:
rnd.shuffle(non_covid)

In [63]:
cnt=0
for i in non_covid:
    if(cnt>=df.shape[0]):
        break      
    src=os.path.join(non_covid_path,i)
    dest=os.path.join(non_covid_folder,i)
    shutil.copy2(src,dest)
    cnt+=1
    
    

In [70]:
train_path='Dataset/train'
val_path='Dataset/val'

In [65]:
import numpy as np
import matplotlib.pyplot as plt
from keras.layers import *
from keras.models import *
from keras.preprocessing import image

#CNN MODEL

In [105]:
model=Sequential()
model.add(Conv2D(32,(3,3),activation='relu',input_shape=(224,224,3)))
model.add(Conv2D(64,(3,3),activation='relu',input_shape=(224,224,3)))
model.add(MaxPool2D((2,2)))
model.add(Dropout(0.25))
model.add(Conv2D(64,(3,3),activation='relu',input_shape=(224,224,3)))
model.add(MaxPool2D((2,2)))
model.add(Dropout(0.25))
model.add(Conv2D(128,(3,3),activation='relu',input_shape=(224,224,3)))
model.add(MaxPool2D((2,2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(64,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(1,activation='sigmoid'))
model.summary()

Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_24 (Conv2D)           (None, 222, 222, 32)      896       
_________________________________________________________________
conv2d_25 (Conv2D)           (None, 220, 220, 64)      18496     
_________________________________________________________________
max_pooling2d_18 (MaxPooling (None, 110, 110, 64)      0         
_________________________________________________________________
dropout_24 (Dropout)         (None, 110, 110, 64)      0         
_________________________________________________________________
conv2d_26 (Conv2D)           (None, 108, 108, 64)      36928     
_________________________________________________________________
max_pooling2d_19 (MaxPooling (None, 54, 54, 64)        0         
_________________________________________________________________
dropout_25 (Dropout)         (None, 54, 54, 64)       

In [106]:
model.compile(loss="binary_crossentropy",optimizer="adam",metrics=["accuracy"])

In [107]:
train_gen=image.ImageDataGenerator(rescale=1.0/255,shear_range=0.2,zoom_range=0.3,horizontal_flip=True)
train_generator=train_gen.flow_from_directory(
train_path,
target_size=(224,224),
batch_size=32,
class_mode='binary'    
)

Found 328 images belonging to 2 classes.


In [108]:
val_gen=image.ImageDataGenerator(rescale=1.0/255)
val_generator=train_gen.flow_from_directory(
val_path,
target_size=(224,224),
batch_size=32,
class_mode='binary'    
)

Found 64 images belonging to 2 classes.


In [109]:
val_generator.class_indices

{'Covid': 0, 'non-covid': 1}

In [110]:
hist2=model.fit(x=train_generator,epochs=8,validation_data=val_generator)

Epoch 1/8
11/11 [==============================] - 50s 4s/step - loss: 1.1651 - accuracy: 0.5305 - val_loss: 0.6885 - val_accuracy: 0.5000
Epoch 2/8
11/11 [==============================] - 51s 5s/step - loss: 0.6478 - accuracy: 0.6341 - val_loss: 0.5821 - val_accuracy: 0.7812
Epoch 3/8
11/11 [==============================] - 50s 4s/step - loss: 0.4644 - accuracy: 0.7622 - val_loss: 0.4244 - val_accuracy: 0.8594
Epoch 4/8
11/11 [==============================] - 47s 4s/step - loss: 0.3943 - accuracy: 0.8171 - val_loss: 0.2766 - val_accuracy: 0.9062
Epoch 5/8
11/11 [==============================] - 59s 5s/step - loss: 0.3129 - accuracy: 0.8841 - val_loss: 0.2552 - val_accuracy: 0.9219
Epoch 6/8
11/11 [==============================] - 50s 4s/step - loss: 0.2875 - accuracy: 0.8841 - val_loss: 0.1906 - val_accuracy: 0.9375
Epoch 7/8
11/11 [==============================] - 49s 5s/step - loss: 0.2732 - accuracy: 0.8872 - val_loss: 0.1969 - val_accuracy: 0.9219
Epoch 8/8
11/11 [==========

In [111]:
model.evaluate(train_generator)

11/11 [==============================] - 16s 1s/step - loss: 0.1488 - accuracy: 0.9665


[0.14876621961593628, 0.9664633870124817]

In [112]:
model.evaluate(val_generator)

2/2 [==============================] - 3s 1s/step - loss: 0.0864 - accuracy: 1.0000


[0.08640404045581818, 1.0]

#TRAINING ACCURACY WE GOT IS ALMOST 97% WITH A VALIDATION SCORE OF 100%!!